quinn-proto/src/connection/streams/recv.rs

```diff
@@ -257,71 +198,179 @@ impl Recv {
     }
 }
 
-pub(crate) type ReadResult<T> = Result<Option<DidRead<T>>, ReadError>;
-
-/// Result of a `Streams::read` call in case the stream had not ended yet
-#[derive(Debug, Eq, PartialEq, Copy, Clone)]
-#[must_use = "A frame might need to be enqueued"]
-pub(crate) struct DidRead<T> {
-    pub result: T,
-    pub max_stream_data: ShouldTransmit,
-    pub max_data: ShouldTransmit,
+/// An iterator-like interface to read data from a receive stream
+pub struct Chunks<'a> {
+    id: StreamId,
+    streams: &'a mut Streams,
+    pending: &'a mut Retransmits,
+    recv: Option<Recv>,
+    ordered: bool,
+    read: u64,
+    fused: Option<ReadStatus>,
 }
 
-pub(super) type StreamReadResult<T> = Result<Option<T>, ReadError>;
+impl<'a> Chunks<'a> {
+    pub(crate) fn new(
+        id: StreamId,
+        streams: &'a mut Streams,
+        pending: &'a mut Retransmits,
+        ordered: bool,
+    ) -> Result<Self, ReadableError> {
+        let entry = match streams.recv.entry(id) {
+            Entry::Occupied(entry) => entry,
+            Entry::Vacant(_) => return Err(ReadableError::UnknownStream),
+        };
 
-pub(crate) trait BytesRead {
-    fn bytes_read(&self) -> u64;
-}
+        let mut rs = match entry.get().stopped {
+            true => return Err(ReadableError::UnknownStream),
+            false => entry.remove(),
+        };
 
-impl BytesRead for Chunk {
-    fn bytes_read(&self) -> u64 {
-        self.bytes.len() as u64
+        rs.assembler.ensure_ordering(ordered)?;
+        Ok(Self {
+            id,
+            recv: Some(rs),
+            streams,
+            pending,
+            ordered,
+            read: 0,
+            fused: None,
+        })
     }
-}
 
-pub(crate) struct ReadChunks {
-    pub bufs: usize,
-    pub read: usize,
+    /// Read the next chunk from the stream
+    ///
+    /// Return the next chunk of data from the stream, if any are available. `max_length` may be
+    /// used to restrict the maximum size of the returned `Bytes` (but `usize::MAX` is supported,
+    /// too).
+    pub fn next(&mut self, max_length: usize) -> ReadResult {
+        if let Some(status) = &self.fused {
+            return ReadResult::from_status(*status);
+        }
+
+        let rs = self.recv.as_mut().unwrap();
+        let chunk = rs.assembler.read(max_length, self.ordered);
+        let status = match &chunk {
+            Some(chunk) => {
+                self.read += chunk.bytes.len() as u64;
+                ReadStatus::Readable
+            }
+            None => match rs.state {
+                RecvState::Recv { size } => {
+                    if size == Some(rs.end) && rs.assembler.bytes_read() == rs.end {
+                        self.streams.stream_freed(self.id, StreamHalf::Recv);
+                        self.fused = Some(ReadStatus::Finished);
```

Comment:
Maybe assign `fused` after the match block, for the `status != ReadStatus::Readable` case

Reply:
A more descriptive name than `fused` could also be nice. I know it resembles the iterator stuff. But for someone not familiar with those APIs it just sounds like a weird word.

Reply:
I've replaced this with a `State` enum.